scipy/sparse/linalg/propack.py

```diff
@@ -0,0 +1,329 @@
+"""
+Python wrapper for PROPACK
+--------------------------
+
+PROPACK is a collection of Fortran routines for iterative computation
+of partial SVDs of large matrices or linear operators.
+
+Based on BSD licensed pypropack project:
+  http://github.com/jakevdp/pypropack
+  Author: Jake Vanderplas <vanderplas@astro.washington.edu>
+
+PROPACK source is BSD licensed, and available at
+  http://soi.stanford.edu/~rmunk/PROPACK/
+"""
+
+__all__ = ['svdp']
+
+import numpy as np
+from scipy.sparse.linalg import aslinearoperator
+from scipy.linalg import LinAlgError
+
+from ._propack import _spropack
+from ._propack import _dpropack
+from ._propack import _cpropack
+from ._propack import _zpropack
+
+
+_lansvd_dict = {
+    'f': _spropack.slansvd,
+    'd': _dpropack.dlansvd,
+    'F': _cpropack.clansvd,
+    'D': _zpropack.zlansvd,
+}
+
+
+_lansvd_irl_dict = {
+    'f': _spropack.slansvd_irl,
+    'd': _dpropack.dlansvd_irl,
+    'F': _cpropack.clansvd_irl,
+    'D': _zpropack.zlansvd_irl,
+}
+
+_which_converter = {
+    'LM': 'L',
+    'SM': 'S',
+}
+
+class _AProd:
+    """
+    Wrapper class for linear operator
+
+    The call signature of the __call__ method matches the callback of
+    the PROPACK routines.
+    """
+    def __init__(self, A):
+        try:
+            self.A = aslinearoperator(A)
+        except TypeError:
+            self.A = aslinearoperator(np.asarray(A))
+
+    def __call__(self, transa, m, n, x, y, sparm, iparm):
+        if transa == 'n':
+            y[:] = self.A.matvec(x)
+        else:
+            y[:] = self.A.rmatvec(x)
+
+    @property
+    def shape(self):
+        return self.A.shape
+
+    @property
+    def dtype(self):
+        try:
+            return self.A.dtype
+        except AttributeError:
+            return self.A.matvec(np.zeros(self.A.shape[1])).dtype
+
+
+def svdp(A, k, which='LM', irl_mode=False, kmax=None,
```

Comment:
```suggestion
def svdp(A, k, which='LM', irl_mode=True, kmax=None,
```
A brief test suggests that `irl_mode=False` is typically faster but less robust.
```
Running benchmarks for Scipy version 1.8.0.dev0+1432.3c23240 at C:\Users\matth\scipy\build\testenv\Lib\site-packages\scipy\__init__.py
· No `environment_type` specified in asv.conf.json. This will be required in the future.
· Discovering benchmarks
· Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
[  0.00%] ·· Benchmarking existing-pyc__users_matth_anaconda3_envs_scipydev_python.exe
[ 50.00%] ··· Running (sparse_linalg_svds.BenchSVDS.time_svds--).
[100.00%] ··· sparse_linalg_svds.BenchSVDS.time_svds                                                                                                                        6/60 failed
[100.00%] ··· ===== ========== ================================= ==================================
              --                                        solver_and_options
              ---------------- --------------------------------------------------------------------
                k    problem    ('propack', {'irl_mode': True})   ('propack', {'irl_mode': False})
              ===== ========== ================================= ==================================
                20    abb313              4.70±0.01ms                       3.67±0.04ms
                20   illc1033              18.0±0.1ms                        6.89±0.2ms
                20   illc1850              21.4±0.1ms                          failed
                20    qh1484               17.6±0.2ms                       6.32±0.07ms
                20   rbs480a              15.9±0.07ms                          failed
                20   tols4000              68.0±0.6ms                          failed
                20   well1033             7.09±0.06ms                       6.27±0.03ms
                20   well1850              14.8±0.2ms                          failed
                20   west0479              8.16±0.3ms                       2.36±0.01ms
                20   west2021              16.0±0.1ms                        7.41±0.1ms
                50    abb313              8.25±0.08ms                       7.81±0.07ms
                50   illc1033              30.9±0.3ms                        32.8±0.1ms
                50   illc1850              62.9±0.5ms                          failed
                50    qh1484               102±0.4ms                         18.7±0.1ms
                50   rbs480a               29.2±0.2ms                        25.2±0.1ms
                50   tols4000               347±1ms                          49.8±0.3ms
                50   well1033             55.4±0.07ms                       22.3±0.07ms
                50   well1850              89.3±0.3ms                          failed
                50   west0479              35.3±0.2ms                       8.70±0.04ms
                50   west2021              104±0.4ms                         17.5±0.2ms
               100    abb313              8.25±0.03ms                       8.20±0.03ms
               100   illc1033              81.0±0.4ms                        61.4±0.2ms
               100   illc1850              104±0.5ms                         75.9±0.4ms
               100    qh1484                428±5ms                          69.0±0.1ms
               100   rbs480a               93.4±0.5ms                        34.7±0.1ms
               100   tols4000               235±1ms                           90.0±3ms
               100   well1033              108±0.4ms                         58.2±0.6ms
               100   well1850               99.1±1ms                         112±0.7ms
               100   west0479              126±0.2ms                         25.1±0.1ms
               100   west2021               432±2ms                           70.2±3ms
              ===== ========== ================================= ==================================
```
I think robustness should be prioritized over speed for basic use.

Since we're only exposing PROPACK via `svds` for now, maybe it's OK to leave this alone here, though. Up to you.

Reply:
Is this worth having a note in the docs?  This is an quick, informal test but seems like it would be useful information for an end-user to know

Reply:
My vote is to say nothing and leave this as the default